File: Programers/Level2/게임맵최단거리.py
from collections import deque

def solution(maps):
    
    dx = [0,0,1,-1]
    dy = [1,-1,0,0]
    
    row = len(maps)
    col = len(maps[0])
    
    visited = [ [-1 for i in range(col)] for _ in range(row)]    
    
    q = deque()
    q.append([0,0])
    visited[0][0] = 1
    
    while q :
        
        y, x = q.popleft()
        
        for i in range(4) :
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0 <= nx < col and 0 <= ny < row :
                if maps[ny][nx] == 1 and visited[ny][nx] == -1 :
                    visited[ny][nx] = visited[y][x] + 1
                    q.append([ny,nx])
        
    answer = visited[-1][-1]
    
    return answer

File: Javascript/Programmers/게임맵최단거리.js
function solution(maps) {
    let answer = 1;

    const dx = [1, -1, 0, 0];
    const dy = [0, 0, 1, -1];
    const row = maps.length;
    const col = maps[0].length;
    let q = [];
    let visited = maps;

    visited[0][0] = 0;
    q.push([0, 0]);

    while (q.length > 0) {
        let qLen = q.length;
        for (let i = 0; i < qLen; i++) {
            let [y, x] = q.shift();

            for (let j = 0; j < 4; j++) {
                let nx = x + dx[j];
                let ny = y + dy[j];

                if (
                    0 <= nx &&
                    nx < col &&
                    0 <= ny &&
                    ny < row &&
                    visited[ny][nx] === 1
                ) {
                    if (nx === col - 1 && ny === row - 1) {
                        return ++answer;
                    }
                    q.push([ny, nx]);
                    visited[ny][nx] = 0;
                }
            }
        }
        
        answer++;
    }
    return -1;
}
